### DISCUSSION TITLE
Vector algebra of graphics?

### DISCUSSION THREAD

**Discussion by mjskay** [ggplot2-extenders/ggplot-extension-club#66] [General]
I posted this thought [on bluesky](https://bsky.app/profile/mjskay.com/post/3lcigw6bumc2v) but this might be a better place:

If a data vis API has ops for superimposing layers (`+`), combining layer properties (`*`), and juxtaposing sums of layers (vector concatentation: `c()`), then the matrix product of a column vector of y encodings (`aes(y = var1...n)`) and a row vector of x encodings (`aes(x = var1...n)`) is a scatterplot matrix.

(I'm not sure if you can do this in Julia's Algebra of Graphics? A quick Google was not fruitful...)

More generally, imagine what you could do with an API built around vectors of sums of layers and vectors of transformation functions on layers. Seems like one way to generalize what {ggblend}, {ggbuilder}, and {patchwork} all kind of get at, but don't fully formalize. I might try to take {ggbuilder} in that direction and see what happens...

**Comment by teunbrand** [comment]
In case anybody else also had to google these:

**superimpose**
To put one thing on top of something else

**juxtaposition**
Putting things that are not similar next to each other

But what do you mean with 'sums of layers' and what would it look like if they are put next to each other? I'm afraid I don't quite follow along with your description.

With regards to scatterplot matrices, I think the following discussion declared it out of scope for vanilla ggplot2: https://github.com/tidyverse/ggplot2/issues/3357
That shouldn't stop anyone from playing around with it though.

**Reply by yjunechoe** [reply]
BTW how early does the Facet touch the plot data? Could you have a decorator like `pack()` and a Facet like `facet_unpack()` such that given a syntax like this:

```r
ggplot(aes(x = pack(x1, x2, x3), y = pack(y1, y2, y3))) + 
  geom_point() +
  facet_unpack()
```

To get back a scatterplot matrix like:

```r
outer(1:3, 1:3, sprintf, fmt = "x%i-y%i")
#>      [,1]    [,2]    [,3]   
#> [1,] "x1-y1" "x1-y2" "x1-y3"
#> [2,] "x2-y1" "x2-y2" "x2-y3"
#> [3,] "x3-y1" "x3-y2" "x3-y3"
```

Where I'd imagine `pack()` to ensure equal length of its vector elements and `facet_unpack()` to act early in `ggplot_build()` to e.g., turn the data long and do a `facet_grid()`-esque `PANEL` assignment for the rest of the build process.

---

Also, saving this Thomas quote from that issue for later :D

> To be clear, I think it is very important to acknowledge that ggplot2 is not an implementation of Wilkinson’s grammar of graphics but an adaption with specific deviations to make it more amenable to a consistent programming API. I think that it is very important not to loose that of sight when we consider adding these sort of things.
>
> This is not to detract from the usefulness of such a feature — it sure would be nice — but not at the cost of major complications in API or implementation

**Reply by teunbrand** [reply]
Facets touch the data for the first time very early, even before aesthetics are evaluated from user data. To be exact, this line of the build method lets facets insert their `PANEL` variable in the user data: https://github.com/tidyverse/ggplot2/blob/73b41194b036776a12d3ec7d5adaeda5d4f81a33/R/plot-build.R#L59.

> To get back a scatterplot matrix like:

No kronecker sprintf this time? :(

**Reply by mjskay** [reply]
Oh for sure, I'm not proposing changing base ggplot2, this is a radical change and more suitable for an extension.

Also apologies, I should have been more clear (I condensed the original post for bluesky's character limit).

"Superimpose" is what vis academics use to just mean "plotting on top of"; i.e. it's what ggplot's `+` operation already does when you add two layers together. So by "sums of layers" I mean "superimposed layers"; i.e. what you get out of summing layers in ggplot currently.

"Juxtapose" is essentially what faceting does, but might be considered a more elementary operation on which you could build faceting. 

So the idea is you could have three things that can be plotted:

1. A layer
2. A layer sum (plotted as its layers superimposed)
3. A vector of layer sums (plotted as each layer sum next to each other, a la faceting / small multiples)

If you also allow vectors of layer operations (things like setting aesthetics, scales, etc) and define `<layer operation> * <layer>` such that it applies the layer operation to the layer (e.g. `aes(x = var) * geom_point()` yields `geom_point(aes(x = var))`), then you can construct a scatterplot matrix via something like (fake notation):

```r
e_x = [aes(x = var[i]) : i ∈ 1..n]
e_y = [aes(y = var[i]) : i ∈ 1..n]
plot = (e_x' × e_y) * geom_point()
```

And even set upper/lower/on-diagonal plots differently:

```r
lower_tri(plot) *= geom_point()
upper_tri(plot) *= geom_heatmap()
diag(plot) *= geom_density() * aes(y = NULL)
```

**Reply by yjunechoe** [reply]
> even before aesthetics are evaluated from user data.

Nice!

> No kronecker sprintf this time? :(

The one time when I actually needed to return a matrix to show my point :')

**Reply by yjunechoe** [reply]
> lower_tri(plot) *= geom_point()
> upper_tri(plot) *= geom_heatmap()
> diag(plot) *= geom_density() * aes(y = NULL)

Also that looks super slick! (Tangentially related - I hate that R has `diag<-` but not `upper.tri<-` and `lower.tri<-` 🥴

**Reply by teunbrand** [reply]
>  I hate that R has diag<- but not upper.tri<- and lower.tri<-

We should also have `head<-` and `tail<-` too

**Reply by teunbrand** [reply]
But yeah that looks like a slick API for scatterplot matrices!
Would one also be able to do the following:

```
plot[1, ] *= annotate()
```
To tailor annotations and such?

**Reply by mjskay** [reply]
```r
plot[1, ] *= annotate()
```

Right exactly! Expose all the great stuff R already has for vector/matrix algebra. Easy to do this in S7 these days :). And you wouldn't need to limit to scatterplot matrices, though it's a clear initial use case.

Depending on the implementation I think it could also address the problem of sharing (or not) scales across facets: if you wanted every subplot to have the same scale for an aesthetic, you multiply the whole thing by one scale; if you want "free" scales you multiply individual plots by their own scales, or if you want columns to share scales you multiply columns by a scale, etc.

**Reply by teunbrand** [reply]
I've already build an `at_panel()` function once. It isn't too hard to express `diag()`, `upper.tri()` and `lower.tri()` as comparisons between row and column numbers.

``` r
library(ggplot2)

ggplot(diamonds, aes(depth, price)) +
  geom_point() +
  facet_grid(color ~ clarity) +
  ggh4x::at_panel(
    annotate("point", I(0.5), I(0.5), colour = "red", size = 5),
    ROW == COL
  ) +
  ggh4x::at_panel(
    annotate("point", I(0.5), I(0.5), colour = "blue", size = 5),
    ROW > COL
  ) +
  ggh4x::at_panel(
    annotate("point", I(0.5), I(0.5), colour = "green", size = 5),
    ROW < COL
  )
```

![](https://i.imgur.com/s0la8bX.png)

<sup>Created on 2024-12-04 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by mjskay** [reply]
Nice!  Could have been useful to past me :)

**Comment by yjunechoe** [comment]
Starting a new comment thread re: Julia AlgebraOfGraphics --

I don't think this translates to a scatterplot matrix (facetting is handled by `row` and `col` mappings in ex: `mapping(:x, :y, row=:i, col=:j)`)

But I can at least see that the dot-product behavior between aesthetics seems to exist in some form (though I have no idea what the result actually means 😆):

```r
(mapping(x = :x1) + mapping(x = :x2)) * (mapping(y = :y1) + mapping(y = :y2))
#> Layers(
#>   Layer[
#>     Layer(identity, nothing, Any[], {:x = :x1, :y = :y1}),
#>     Layer(identity, nothing, Any[], {:x = :x1, :y = :y2}),
#>     Layer(identity, nothing, Any[], {:x = :x2, :y = :y1}),
#>     Layer(identity, nothing, Any[], {:x = :x2, :y = :y2})
#>   ]
#> )
```

**Reply by mjskay** [reply]
Heh cool!! If I'm reading that right, it's the superimposition of four layers, one for each combination of the variables?

**Reply by yjunechoe** [reply]
I'll let you know once Makie finishes precompiling... 😭

**Reply by yjunechoe** [reply]
Ah! it just gives you 4 plots! Interesting... (or maybe I specified this incorrectly -- much of this API is foreign and going straight over my head 😅)

```julia
using AlgebraOfGraphics
using DataFrames
using PalmerPenguins
penguins = dropmissing(DataFrame(PalmerPenguins.load()));
plt = data(penguins) *
  (
    (mapping(:bill_length_mm) + mapping(:bill_depth_mm)) *
    (mapping(:flipper_length_mm) + mapping(:body_mass_g))
  )
length(plt) # 4
draw(plt[1])
```

![display](https://github.com/user-attachments/assets/3167b138-16cc-4ae6-8a84-cd7bae1ce33e)

I think generalizing this to a matrix might end up going the JavaScript (ex: D3) route of facetting, where facets are handled at the point of "rendering" (IIRC in D3, the graphing libs render the actual svgs, and then it's your job afterwards to lay them out appropriately using ex: css grid).

So back to Julia, the 4 plots in `plt` could probably be thrown together into a matrix by `draw()`, which seems to be pretty rich with arguments.

**Reply by mjskay** [reply]
Dang just reminds me "play with Julia" is perpetually on my TODO list...

**Reply by yjunechoe** [reply]
Oh wait I see: it's 4 layers (length = layers), which can each be drawn out individually (which is what i did above). If you `draw(plt)` the 4 layers simultaneously, you get the expected nonsensical plot:

![display](https://github.com/user-attachments/assets/0848997d-25ce-424a-8bc3-2eadbf1cc696)

So I think the facet matrix behavior would still most straightforwardly come from tweaking arguments in `draw()` and treating each element of `plt` as separate in some way

**Reply by mjskay** [reply]
Hm... instead of summing the mappings before multiplying, can you create two length-2 vectors of mappings (or is that what `+` is doing?) and multiply those? (maybe need to transpose one first to get the outer product...)

**Reply by yjunechoe** [reply]
> Dang just reminds me "play with Julia" is perpetually on my TODO list...

Please do cool stuff in Julia! Plotting is getting more painless!

![image](https://github.com/user-attachments/assets/a90e6e78-6730-4578-8a7b-8ec2f76123c2)

https://www.youtube.com/watch?v=sXhRWCV38iQ

**Reply by yjunechoe** [reply]
> can you create two length-2 vectors of mappings

I tried that first too (this one seems more intuitive to me!) but it didn't want to cross the elements:

```julia
mapping([:bill_length_mm, :bill_depth_mm]) * mapping([:flipper_length_mm, :flipper_length_mm])
#> Layer(identity, nothing, Any[[:bill_length_mm, :bill_depth_mm], [:flipper_length_mm, :flipper_length_mm]], {})
```

So that just returns a single plot, drawing ~2 columns worth of data at x and y each~ using just the first columns in each mapping (I think.. but definitely not data from all 4 columns involved):

(so I think this just breaks the grammar. Cool that it lets us break it though)
![display](https://github.com/user-attachments/assets/e4fdb655-a0f0-46ba-a210-13b4cfc88b46)


> (or is that what + is doing?)

So yeah, I think things you need the `+` for the later `*` to interpret that as distributive

**Reply by mjskay** [reply]
(based on some quick googling of "julia outer product", what does something like `[mapping(x = :x1), mapping(x = :x2)]' .* [mapping(y = :y1), mapping(y = :y2)]` do?

(also feel free to tell me "go install Julia if you want to figure this out" ;) )

**Reply by yjunechoe** [reply]
Oh! Ok I see what you meant by "transpose" (literally) now (flew past my head after the "superimpose" and "juxtapose" discussion 😆)

And... Hmm!

So, `[mapping(x = :x1), mapping(x = :x2)]'` doesn't work because `'` for `Vector{Layer}` is not defined, but you can get to the transpose more agnostically with `hcat()`. So - this works:

```julia
hcat(mapping(x = :x1), mapping(x = :x2)) .* vcat(mapping(y = :y1), mapping(y = :y2))
#> 2×2 Matrix{Layer}:
#> Layer(identity, nothing, Any[], {:x = :x1, :y = :y1})  Layer(identity, nothing, Any[], {:x = :x2, :y = :y1})
#> Layer(identity, nothing, Any[], {:x = :x1, :y = :y2})  Layer(identity, nothing, Any[], {:x = :x2, :y = :y2})
```

As to what this means -- I just closed out of my other session with makie loaded so I'll get back to this later in the day 😅 (but yes do install Julia to figure this out immediately 😛)

**Reply by mjskay** [reply]
hahah thanks! sadly I have a bunch of other stuff I have to do today so I probably won't get to actually trying stuff until later this week at the earliest :)

**Reply by mjskay** [reply]
(though looking at this and comparing it to your previous output I do wonder if what we're looking for is a matrix of "Layers" objects instead of a matrix of "Layer" objects, if that makes sense...)

**Reply by yjunechoe** [reply]
Right right -- ok so I can take this matrix of `Layer`s:

```julia
m1 = hcat(mapping(:bill_length_mm), mapping(:bill_depth_mm)) .* vcat(mapping(:flipper_length_mm), mapping(:body_mass_g));
m1
#> 2×2 Matrix{Layer}:
#>  Layer(identity, nothing, Any[:bill_length_mm, :flipper_length_mm], {})  …  Layer(identity, nothing, Any[:bill_depth_mm, :flipper_length_mm], {})
#>  Layer(identity, nothing, Any[:bill_length_mm, :body_mass_g], {})           Layer(identity, nothing, Any[:bill_depth_mm, :body_mass_g], {})
```

And get a matrix of `Layers`s:

```julia
m2 = map(x -> Layers([x]), m1);
m2
#> 2×2 Matrix{Layers}:
#>  Layers(Layer[Layer(identity, nothing, Any[:bill_length_mm, :flipper_length_mm], {})])  …  Layers(Layer[Layer(identity, nothing, Any[:bill_depth_mm, :flipper_length_mm], {})])
#>  Layers(Layer[Layer(identity, nothing, Any[:bill_length_mm, :body_mass_g], {})])           Layers(Layer[Layer(identity, nothing, Any[:bill_depth_mm, :body_mass_g], {})])
```

And each element of this matrix can combine with the data to make a fully saturated layer that can actually materialize:

```julia
plt_m2_1 = data(penguins) * m2[1,1];

dump(plt_m2_1; maxdepth=4);
#> Layers
#>   layers: Array{Layer}((1,))
#>     1: Layer
#>       transformation: identity (function of type typeof(identity))
#>       data: DataFrames.DataFrameColumns{DataFrame}
#>         df: DataFrame
#>       positional: Array{Any}((2,))
#>         1: Symbol bill_length_mm
#>         2: Symbol flipper_length_mm
#>       named: Dictionaries.Dictionary{Symbol, Any}
#>         indices: Dictionaries.Indices{Symbol}
#>         values: Array{Any}((0,))

draw(plt_m2_1) # BTW just `draw()`-ing the inner `Layer` (i.e., `plt_m2_1[1]`) works too
```

![display](https://github.com/user-attachments/assets/717ecd26-0f59-4f0a-803a-48bd1072d8b8)

And the syntax we're hoping from this grammar for a scatterplot matrix is... what exactly? 😅 Would it be expressed as a `*` method involving a matrix of Layers?

```julia
data(penguins) * m2
#> ERROR: MethodError: no method matching *(::Layer, ::Matrix{Layers})
```

And/or a `draw()` method for a matrix of `Layers`, such as:

```julia
m2 .* data(penguins) # This line errors and IDK why I can't broadcast here TBH, but imagine element-wise `*`, preserving matrix structure
  |> draw
#> ERROR: MethodError: no method matching draw(::Matrix{Layers})
```

**Reply by mjskay** [reply]
Ah, okay... So in this api a single `Layers` object is what I called a layer sum, so the initial row and column vector should each be a `Layers` vector of length 2 instead of a `Layer` vector of length 2. Then their outer product should be the `Layers` matrix you produced, after which yeah you need methods like `*` and `draw`.

**Reply by yjunechoe** [reply]
Makes sense. And yeah I think AoG just hasn't explored such "geometric" (matrix/rectangle) arrangements of `Layers`:

```julia
methods(*, AlgebraOfGraphics)
#> # 2 methods for generic function "*" from Base:
#>  [1] *(l::Layer, l′::Layer)
#>  [2] *(a::AlgebraOfGraphics.AbstractAlgebraic, a′::AlgebraOfGraphics.AbstractAlgebraic)

supertype(Layer)
#> AlgebraOfGraphics.AbstractAlgebraic

supertype(Layers)
#> AlgebraOfGraphics.AbstractAlgebraic
```

I wonder if there's anything prohibitive (technically or cognitively or otherwise) that makes vis APIs keep reinventing facet row/col as a distinct construct in some form 🤔

**Reply by mjskay** [reply]
Yeah, from the cognitive perspective (or perhaps a task-oriented perspective) we encountered an issue related to this in a study of ggplot2 users awhile back:

<img src="https://github.com/user-attachments/assets/509c8f9e-44f8-40c6-a364-eb5b2e4fbae9" width="500">
<img src="https://github.com/user-attachments/assets/3991e479-c2c4-4267-a94b-5ada20b3a8d5" width="500">
<img src="https://github.com/user-attachments/assets/e6c290fa-e692-4dcc-82ed-d924d5147254" width="500">

Paper [here](https://doi.org/10.1145/3544548.3580837).

**Reply by EvaMaeRey** [reply]
I haven't been keeping up with the conversation, but I think it's at least tangentially relevant:  I remember learning Tableau as a ggplot2 user thinking 'where is the facet declaration field'? But at least at that time, that was not how it was handled.   It's a little disorienting to start Tableau, as there's no x and y fields/declaration - position is all columns (x) and rows (y).  Point of entry thinking about a cross tablulation instead of a scatterplot (which I think is ggplot2's point of entry).  If you want to break up data by categories you can do two col (x) variables:  cols [var_categorical, var_continuous] .  

I do think of faceting differently because of tableau -- as a positional aesthetic.  (but I wasn't really thinking about Stats back then and compute_group v. compute_panel, compute_layer)

![image](https://github.com/user-attachments/assets/1000b2c0-a68c-4d02-b6f5-ff95c82d295b) from https://evolytics.com/blog/tableau-201-how-to-make-small-multiples/

I think wrapping by category was a bit awkward a the time, but maybe that has been addressed.  

But I think most ggplot2 users don't think this way, instead:  'single plot to many plots' thinking (rather than spatial offset/positional aesthetic) probably dominates, and seems like a pretty helpful way of understanding. ("Faceting allows you to turbocharge your graphs so you can go from a single plot to many plots, helping you see lots of patterns." quote from: https://www.linkedin.com/posts/c-jarvis_i-was-teaching-last-night-on-our-shiny-course-activity-7270401305729290241-EmwI?utm_source=share&utm_medium=member_desktop

Please feel free to get back to Julia discussion! :-)

**Reply by mjskay** [reply]
@EvaMaeRey I think you hit it exactly! The question is: is the mental model "I get an extra positional aesthetic for rows and columns" or is it "I am repeating this chart out into small multiples".

**Reply by EvaMaeRey** [reply]
🙏Your paper looks so interesting. 🛠📊 Excited to have a closer look!

**Reply by EvaMaeRey** [reply]
I think a StatFacet (or a go-to computation that could be added as a last step of other compute to define a Stat - w/ stat chaining could be really amazing too) could be useful.  In some areas, like calendars for example, facet_wrap() by month seems like a go-to.  But it feels like a messy solution (you have x and y defined by day in the month - day of week, week in month), but then then the month position is handled in this orthogonal way: facet_wrap(~month(date)).  It seems like you shouldn't have to necessarily do that spatial offset using the faceting mechanism, but could define it in the Stat.   Does that make sense?  I'd be curious about projects that take this on, if you all know of some.

**Reply by EvaMaeRey** [reply]
I had a quick look at writing position_wrap(), given some of the cool things happening w/ position_*() function w/ ggstats and waterfall chart, and elsewhere... But right now it's mostly giving me a greater appreciation for what the facet_ functions do...  

Combining a stat that computes the xy  mean with position_wrap for example doesn't produce a very meaningful plot (which now seems obvious, but I had to kind of experience it to come to feel the obviousness).  Still position_wrap might be useful in some setting like that calendar case (in the case of layer-only work -- but looking forward to possible coord), of some other more grid-type applications.  


``` r
library(tidyverse)
compute_layer_wrap <- function(data, params, panel) {
    
    wrap <- as.numeric(as.factor(data$wrap))
    
    wrapping_x <- (wrap %% 3)
    wrapping_y <- ((wrap-1) %/% 3)
    
    range_x <- (max(data$x)-min(data$x))
    range_y <- (max(data$y)-min(data$y))
    if(range_x == 0){range_x <- 1}
    if(range_y == 0){range_y <- 1}
    
    ggplot2::transform_position(
      df = data,
      trans_x = function(x) {(x - min(x))/(range_x) + wrapping_x}, 
      trans_y = function(y) {(y - max(y))/(range_y) - wrapping_y}
    )
  }

PositionWrap <- ggproto(`_class` = 'PositionWrap', `_inherit` = Position,
                        required_aes = c('x', 'y', 'wrap'),
                        compute_layer = compute_layer_wrap)


position_wrap <- function(sd_x = .15, sd_y = .15) {
  ggproto(NULL, PositionWrap, sd_x = sd_x, sd_y = sd_y)
}

df <- data.frame(
  x = rnorm(50),
  y = rnorm(50),
  index = 1:50
)

library(statexpress)

compute_group_means <- function(data, scales){
  
  data |> 
    summarise(x = mean(x),
              y = mean(y))
  
}

compute_group_label <- function(data, scales){
  
  data |> 
    group_by(label) |>
    summarise(x = 0,
              y = 0)
  
}

ggplot(mtcars) +
  aes(wt, mpg, 
      color = factor(cyl), 
      label = factor(cyl),
      wrap = interaction(gear, cyl)) +
  geom_point(position = position_wrap()) + 
  geom_point(stat = ggproto("temp", Stat, 
                            compute_group = compute_group_means), 
             position = position_wrap(),
             size = 8, alpha = .2) +
  geom_label(stat = ggproto("temp", Stat, 
                            compute_group = compute_group_label),
             position = position_wrap(),
             vjust = 1, hjust = 0 ) +
  geom_vline(xintercept = 0:3, linetype = "dashed") + 
  geom_hline(yintercept = 0:-3, linetype = "dashed")
```

![image](https://github.com/user-attachments/assets/dc143e26-8f77-469e-a71f-bf0d52a0e0cf)

**Reply by teunbrand** [reply]
>  a greater appreciation for what the facet_ functions do

Arguebly, facets are the most complicated parts about a plot 😅

**Reply by EvaMaeRey** [reply]
And they feel so easy/obvious to use!

**Reply by EvaMaeRey** [reply]
position_wrap in gguke, nice behavior (as far as I can tell) in a grid setting!  https://github.com/EvaMaeRey/gguke?tab=readme-ov-file#then-we-need-a-fret-board

**Reply by EvaMaeRey** [reply]
Not sure if this is being done, but using the newly working asis function `I` w/ (x) and (y) in position function might be cool (in the example the positioning and scaling is kind of hard coded - proof of concept!)  I think there are a bunch of tools dedicated to insets and these kind of arbitrary placements, but I'm not sure how they work under the hood.  ggmagnify is one I've seen recently - I think it doesn't use I() but not sure. 

```r
library(tidyverse)
theme_set(theme_bw())

compute_layer_applique <- function(data, params, panel, ncol = 3, x0 = .05, y0 = .75, x_scale = .2, y_scale = .2) {
    
    range_x <- (max(data$x)-min(data$x))
    range_y <- (max(data$y)-min(data$y))
    if(range_x == 0){range_x <- 1}
    if(range_y == 0){range_y <- 1}
    
    ggplot2::transform_position(
      df = data,
      trans_x = function(x) {I(x_scale * x/(range_x) + x0)}, 
      trans_y = function(y) {I(y_scale * y/(range_y) + y0)}
    )
  }

PositionApplique <- ggproto(`_class` = 'PositionApplique', 
                        `_inherit` = Position,
                        required_aes = c('x', 'y'),
                        compute_layer = compute_layer_applique)


position_applique <- function() {
  ggproto(NULL, PositionApplique)
}

library(statexpress)

ggplot(cars) + 
  aes(x = speed) + 
  geom_histogram() + 
  geom_histogram(position = "applique", 
                                fill = "darkred")
```

![image](https://github.com/user-attachments/assets/5dade19d-464c-4afb-b640-b3cc6ecbe77c)


``` r
ggplot(cars) + 
  aes(speed) + 
  geom_point(aes(y = dist)) + 
  geom_histogram(position = "applique", 
                 fill = "darkred")
```


![image](https://github.com/user-attachments/assets/a1984a5c-86e0-452c-92cb-113bbd17d9d7)
